Docker more info (#575)
huggingface/hub-docs: docs/hub/spaces-sdks-docker.md

# Docker Spaces

> Spaces accommodate custom [Docker containers](https://docs.docker.com/get-started/) for apps outside the scope of Streamlit and Gradio. Docker Spaces allows users to go beyond the limits of what was previously possible with the standard sdks. From FastAPI and Go endpoints to Phoenix apps and ML Ops tools, Docker Spaces can help in many different setups.

Reviewer: ~~maybe let's revert this to keep this PR focused on the main changes?~~ no big deal

> Spaces accommodate custom [Docker containers](https://docs.docker.com/get-started/) for apps outside the scope of Streamlit and Gradio. Docker Spaces allows users to go beyond the limits of what was previously possible with the standard sdks. From FastAPI and Go endpoints to Phoenix apps and ML Ops tools, Docker Spaces can help in many different setups.

Reviewer 2: ```suggestion
Spaces accommodate custom [Docker containers](https://docs.docker.com/get-started/) for apps outside the scope of Streamlit and Gradio. Docker Spaces allow users to go beyond the limits of what was previously possible with the standard SDKs. From FastAPI and Go endpoints to Phoenix apps and ML Ops tools, Docker Spaces can help in many different setups.
```

---
```

Internally you could have as many open ports as you want. For instance, you can install Elasticsearch inside your Space and call it internally on its default port 9200.

If you want to expose apps served on multiple ports to the outside world, a workaround is to use a reverse proxy like Nginx to dispatch requests from the broader internet (on a single port) to different internal ports.

## Secret Management

### Buildtime

In Docker Spaces, the secrets management is different for security reasons. Once you create a secret in the [Settings tab](./spaces-overview#managing-secrets), you can expose the secret by adding the following line in your Dockerfile.

> For example, `SECRET_EXAMPLE` is the name of the secret you created in the Settings tab.
> 

Reviewer: ```suggestion
For example, if `SECRET_EXAMPLE` is the name of the secret you created in the Settings tab, you can read it at build time by doing `$(cat /run/secrets/SECRET_EXAMPLE)`.

See an example below:

```

what about also adding just an example like

---
```

> Internally you could have as many open ports as you want - and use a reverse proxy like Nginx to dispatch requests from the broader internet.

Reviewer: ```suggestion
Internally you could have as many open ports as you want, for instance you can install Elasticsearch inside your Space and call it internally on its default port 9200.

If you want to expose apps served on multiple ports to the outside world, a workaround is to use a reverse proxy like Nginx to dispatch requests from the broader internet (on a single port) to different internal ports.
```

---
```

> Internally you could have as many open ports as you want, for instance you can install Elasticsearch inside your Space and call it internally on its default port 9200.

Reviewer 2: ```suggestion
Internally you could have as many open ports as you want. For instance, you can install Elasticsearch inside your Space and call it internally on its default port 9200.
```

COPY --chown=user . $HOME/app
```

## Data Persistence

> We don't provide a way to persist data in the Docker Spaces. However, we suggest using our [Datasets hub](datasets) for specific cases, where you can store state and data in a git LFS repository. In all other cases, you might need an external storage solution DB, S3, etc.

Reviewer: ```suggestion
We don't yet provide a built-in way to persist data in the Docker Spaces. However, we suggest using our [Datasets hub](datasets) for specific cases, where you can store state and data in a git LFS repository. In other cases, you might want to use from your Space's code an external storage solution like an external hosted DB, S3, etc.
```

COPY --chown=user . $HOME/app
```

## Data Persistence

> We don't yet provide a built-in way to persist data in the Docker Spaces. However, we suggest using our [Datasets hub](datasets) for specific cases, where you can store state and data in a git LFS repository. In other cases, you might want to use from your Space's code an external storage solution like an external hosted DB, S3, etc.

Reviewer 2: ```suggestion
We don't yet provide a built-in way to persist data in the Docker Spaces. However, we suggest using our Datasets Hub for specific cases, where you can store state and data in a git LFS repository. You can find an example of persistence [here](https://huggingface.co/spaces/julien-c/persistent-data), which uses the [`huggingface_hub` library](https://huggingface.co/docs/huggingface_hub/index) for programmatically uploading files to a dataset repository.

In other cases, you might want to use an external storage solution from your Space's code like an external hosted DB, S3, etc.
```
